File: ImplementQueueUsingStacks.py
"""
232. Implement Queue using Stacks
Implement the following operations of a queue using stacks.

push(x) -- Push element x to the back of queue.
pop() -- Removes the element from in front of queue.
peek() -- Get the front element.
empty() -- Return whether the queue is empty.
Notes:
You must use only standard operations of a stack -- which means only push to top, peek/pop from top, size, and is empty operations are valid.
Depending on your language, stack may not be supported natively. You may simulate a stack by using a list or deque (double-ended queue), as long as you use only standard operations of a stack.
You may assume that all operations are valid (for example, no pop or peek operations will be called on an empty queue).
"""
class Queue(object):
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []

    def push(self, x):
        """
        :type x: int
        :rtype: nothing
        """
        self.stack.append(x)

    def pop(self):
        """
        :rtype: nothing
        """
        stack = []
        for i in range(len(self.stack) - 1):
            stack.append(self.stack.pop())
        self.stack.pop()
        while stack:
            self.stack.append(stack.pop())

    def peek(self):
        """
        :rtype: int
        """
        return self.stack[0]

    def empty(self):
        """
        :rtype: bool
        """
        return len(self.stack) == 0

File: Java/ImplementQueueUsingStacks.java
import java.util.Stack;


public class ImplementQueueUsingStacks {
    Stack<Integer> stack1 = new Stack<Integer>();
    Stack<Integer> stack2 = new Stack<Integer>();
    
    // Push element x to the back of queue.
    public void push(int x) {
        stack1.push(x);
    }

    // Removes the element from in front of queue.
    public void pop() {
        if(stack2.size() == 0) {
            stack1ToStack2();
        }
        stack2.pop();
    }

    // Get the front element.
    public int peek() {
        if(stack2.size() == 0) {
            stack1ToStack2();
        }
        return stack2.peek();
    }
    
    private void stack1ToStack2() {
        while(stack1.size() != 0) {
            stack2.push(stack1.pop());
        }
    }

    // Return whether the queue is empty.
    public boolean empty() {
        return stack1.isEmpty() && stack2.isEmpty();
    }

}
